### ISSUE TITLE
documentation example: for else

### ISSUE THREAD

**Issue by vishen** [ziglang/zig#2614]
The documentation example for `for else` (https://ziglang.org/documentation/master/#for)  doesn't run the `else` branch of the code:

```zig
test "for else" {
    // For allows an else attached to it, the same as a while loop.
    var items = []?i32 { 3, 4, null, 5 };

    // For loops can also be used as expressions.
    var sum: i32 = 0;
    const result = for (items) |value| {
        if (value == null) {
            break 9;
        } else {
            sum += value.?;
        }
    } else blk: {
        assert(sum == 7);
        break :blk sum;
    };
}
```

I don't know if the `break` is then supposed to cause the `else` block to run?  But currently the `else` block isn't actually run.

I was just wondering what the expected behavior is here?

**Comment by shritesh** [comment]
When you break from a while/for loop, the else branch is not evaluated. I’ve submitted #2641 to clarify that in the docs.

**Comment by vishen** [comment]
Would it also be useful to have the example use the else branch in the documentation? For me, who is new to Zig, it was a bit confusing having an example that didn't use the else branch here (since it `breaks` out when it see's a `null` and therefore never runs the `else` branch).

**Comment by andrewrk** [comment]
> Would it also be useful to have the example use the else branch in the documentation?

Yes, I think so

**Comment by vishen** [comment]
I was able to make the `else` branch run using the following (https://github.com/ziglang/zig/pull/2642):

```zig
test "for else" {
    // For allows an else attached to it, the same as a while loop.
    var items = []?i32{ 3, 4, null, 5 };

    // For loops can also be used as expressions.
    var sum: i32 = 0;
    const result = for (items) |value| {
        if (value != null) {
            sum += value.?;
        }
    } else blk: {
        assert(sum == 12);
        break :blk sum;
    };
    assert(result == 12);
}
```

However, I am wondering if this is the correct behaviour? I am probably using this wrong, but to me it looks like the `else` branch is only run once the for-loop has finished iterating.

**Comment by emekoi** [comment]
if i understand correctly the `else` branch is only run if the loop doesn't break.

**Comment by andrewrk** [comment]
The most helpful way to think of it is like the `else` of an `if` statement, but for a `while` loop. Specifically, for an error union:

```zig
while (getNextItem()) |item| {
    doSomething(item);
} else |err| {
    std.debug.warn("stopped processing items: {}\n", err);
}
```

Without the `else` there would be no way to capture the `err` value, which is the main reason that loops have the `else` clause.

**Comment by daurnimator** [comment]
But a for loop doesn't throw exceptions?

**Comment by vishen** [comment]
I think the following _should_ work as you described, however I get a syntax error; just wondering what I am doing wrong?

```zig
test "for else" {
    // For allows an else attached to it, the same as a while loop.
    var items = []anyerror!i32{ 3, 4, error.NoneAvailable, 5 };

    // For loops can also be used as expressions.
    var sum: i32 = 0;
    for (items) |value| {
        sum += value;
    } else |err| {
        warn("error in items\n");
    }
    assert(sum == 7);
}
```

```
/home/jonathan/z.zig:12:12: error: invalid token: '|'
    } else |err| {
           ^
```

**Comment by vishen** [comment]
Looking at the parser.cpp file, it looks like the for-loop doesn't capture the error payload (yet?): https://github.com/ziglang/zig/blob/master/src/parser.cpp#L423-L438

However, the while-loop does capture the error payload: https://github.com/ziglang/zig/blob/master/src/parser.cpp#L440-L459

**Comment by andrewrk** [comment]
For loops don't capture error values. But the control flow is the same with `else` on all 7 of these:
 * if bool
 * if optional
 * if error union
 * while bool
 * while optional
 * while error union
 * for

my point is that the "while error union" example is the reasoning for `else` existing on loops at all. everything else is for consistency.

**Comment by vishen** [comment]
Ah, sorry, gotcha!

**Comment by benjaminmordaunt** [comment]
I still find the for/else example in the docs (as of 0.9.1) to be a confusing use case, and not really demonstrate the purpose of the else clause in a for loop.

It seems to be evaluating the `else` clause as a kind of useless `finally` block?

**Comment by redmaner** [comment]
I was going through ziglings.org exercises and it provides another good example for using the else block in a loop expression. See the exercise here: https://codeberg.org/ziglings/exercises/src/branch/main/exercises/062_loop_expressions.zig

It basically mentions you can use a while or for loop as an expression, by using `break $value`. The else block is executed when the loop doesn't use `break`. 

TL:DR example:
```
// Given an array of numbers 
const numbers: [4]u8 = .{5, 6, 9, 12};

// Search for a number higher than 10 and assign to `higher_than_ten` variable
var higher_than_ten: ?u8 = for (numbers) | num| {
  if (num > 10) break num;
} else null; // this else block is executed when the for loop runs out of items. The default else block when omitted returns void. 
```
